Update building documentation for Ubuntu 22.04

alvarocurt opened the pull request:
When building the updated crates I encountered an error as shown in Issue#216.
I did my small research and managed to solve it.
Accept the PR so that no one else has this problem

raphaelhetzel commented:
Not sure of we should set that as it is only required for a small set of protoc versions.
How does that behave with more recent versions of protoc?

alvarocurt commented:
As the [documentation stays](https://protobuf.dev/programming-guides/field_presence/#how-to-enable-explicit-presence-in-proto3), this flag was only REQUIRED for releases until [v3.12.0](https://github.com/protocolbuffers/protobuf/releases/tag/v3.12.0) (casually the one that Ubuntu 22.04 LTS uses).
For more modern releases of protobuf, I imagine the flag is just an explicit and useless declaration of Proton3 being in the code

alvarocurt commented:
Adding to this discussion, if the devcontainer is based on fedora, couldn't we also state that fedora 40 is supported for edgeless?

ccicconetti commented:
> Adding to this discussion, if the devcontainer is based on fedora, couldn't we also state that fedora 40 is supported for edgeless?

Supporting one distro "officially" would require at the very least some continuous integration job that makes sure compilation works on that distro.

I think we can adopt a lighter approach for now, and close this issue by adding to the building instructions:

1. a mention that protoc with version > 3.12 is required
2. a list of distributions on which EDGELESS has been compiled (so not officially supported but at least that's an indication)

What do you think @raphaelhetzel @alvarocurt ?

alvarocurt commented:
I'm ok about resigning the efforts on mantaining more distros like fedora. It was just and idea which seemed low effort, but I agree we are already busy enough.
Regarding the second comment, I would add instructions on which protoc version/s to use and how to install it/them, to avoid confusion or further problems with a version being too new.

About the list of EDGELESS distros where it has been compiled, I would also add some preferences. Like listing the options in order of more used to less.
I mean, if you guys tend to prefer devcontainers, I would put that as the go-to solution. If the nix flake has been broken for a while now, that one would be the last.
That's my vision, but my approach to this might be wrong haha @raphaelhetzel @ccicconetti

mor1 commented:
(If the nix flake is broken and needs fixing, I didn't know about it :) but
agree, settling on some solution is better than trying to maintain many or
taking a long time thinking about it.)

On Tue, 22 Oct 2024 at 15:25, Álvaro Curto Merino ***@***.***>
wrote:

> I'm ok about resigning the efforts on mantaining more distros like fedora.
> It was just and idea which seemed low effort, but I agree we are already
> busy enough.
> Regarding the second comment, I would add instructions on which protoc
> version/s to use and how to install it/them, to avoid confusion or further
> problems with a version being too new.
>
> About the list of EDGELESS distros where it has been compiled, I would
> also add some preferences. Like listing the options in order of more used
> to less.
> I mean, if you guys tend to prefer devcontainers, I would put that as the
> go-to solution. If the nix flake has been broken for a while now, that one
> would be the last.
> That's my vision, but my approach to this might be wrong haha
> @raphaelhetzel <https://github.com/raphaelhetzel> @ccicconetti
> <https://github.com/ccicconetti>
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/edgeless-project/edgeless/pull/217#issuecomment-2429437571>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AABI7YCQSLE4VEKZKZSTVV3Z4ZN4FAVCNFSM6AAAAABQMEBKRSVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZDIMRZGQZTONJXGE>
> .
> You are receiving this because you are subscribed to this thread.Message
> ID: ***@***.***>
>


-- 
Richard Mortier
***@***.***

alvarocurt commented:
@mor1 Haha I have no idea, I just saw issue #38 and though that, but i haven't tried. I regret my words if its working.

Anyways, I have removed my changes, updated documentation on ubuntu installation with how to install the latest release and the library, and in the process, updated file `edgeless_orc/src/proxy_redis.rs` in order to fix this warning I always get when compiling

```bash
   Compiling edgeless_orc v0.1.0 (/root/edgeless/edgeless_orc)
warning: this function depends on never type fallback being `()`
  --> edgeless_orc/src/proxy_redis.rs:51:5
   |
51 |     pub fn new(redis_url: &str, flushdb: bool, dataset_settings: Option<crate::EdgelessOrcProxyDatasetSettings>) -> anyhow::Result<Self...
   |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
   |
   = warning: this was previously accepted by the compiler but is being phased out; it will become a hard error in a future release!
   = note: for more information, see issue #123748 <https://github.com/rust-lang/rust/issues/123748>
   = help: specify the types explicitly
note: in edition 2024, the requirement `!: FromRedisValue` will fail
  --> edgeless_orc/src/proxy_redis.rs:66:35
   |
66 |             redis::cmd("FLUSHDB").query(&mut connection)?;
   |                                   ^^^^^
   = note: `#[warn(dependency_on_unit_never_type_fallback)]` on by default
```

ccicconetti commented on edgeless_orc/src/proxy_redis.rs:
```diff
@@ -56,11 +56,11 @@ impl ProxyRedis {
         if flushdb {
             // flush the in-memory database upon construction
-            redis::cmd("FLUSHDB").query(&mut connection)?;
+            let _: () = redis::cmd("FLUSHDB").query(&mut connection)?;
```
I see this fixes the warning you got on newer compiler, but I think it would be more correct to specify the type of the query() method, like this:

```
let _ = redis::cmd("FLUSHDB").query::<String>(&mut connection)?;
```

ccicconetti commented on edgeless_orc/src/proxy_redis.rs:
```diff
@@ -56,11 +56,11 @@ impl ProxyRedis {
 
         // create the connection with the Redis server
-        let mut connection = redis::Client::open(redis_url)?.get_connection()?;
+        let mut connection: redis::Connection = redis::Client::open(redis_url)?.get_connection()?;
```
Is there a specific reason why you forced the return type here?

I tried reverting to the original code and I get no warning with the latest compiler (`rustc 1.82.0 (f6e511eec 2024-10-15)`)

alvarocurt replied:
Not really, it was the first thing I tried when trying to fix the warning.
You are right, it is not needed, I'll revert the change
